shoop_tests/functional/test_discount_taxing.py

```diff
@@ -0,0 +1,544 @@
+from collections import defaultdict
+from decimal import Decimal
+
+import pytest
+from django.test.utils import override_settings
+
+from shoop.core.models import OrderLineType, Tax, TaxClass
+from shoop.core.order_creator import OrderSource
+from shoop.core.pricing import TaxlessPrice
+from shoop.core.taxing import TaxSummary
+from shoop.default_tax.models import TaxRule
+from shoop.testing.factories import (
+    create_product, get_payment_method, get_shipping_method, get_shop
+)
+
+
+def setup_module(module):
+    global settings_overrider
+    settings_overrider = override_settings(
+        SHOOP_CALCULATE_TAXES_AUTOMATICALLY_IF_POSSIBLE=False)
+    settings_overrider.__enter__()
+
+
+def teardown_module(module):
+    settings_overrider.__exit__(None, None, None)
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize("taxes", ['taxful', 'taxless'])
+def test_1prod(taxes):
+    source = create_order_source(
+        prices_include_tax=(taxes == 'taxful'),
+        line_data=[
+            'product: P1      |price: 200.00|qty: 1|discount:  0.00|tax: A',
+            'discount: SALE   |price: -20.00|qty: 1|discount: 20.00'],
+        tax_rates={'A': ['0.25']})
+    source.calculate_taxes()
+
+    assert source.total_price.value == 180
+    assert get_price_by_tax_class(source) == {'TC-A': 200, '': -20}
+
+    if taxes == 'taxful':
+        #    Name    Rate    Base amount     Tax amount         Taxful
+        assert get_pretty_tax_summary(source) == [
+            'Tax-A   0.25  144.000000000   36.000000000  180.000000000']
+        assert get_pretty_line_taxes_of_discount_lines(source) == [[
+            'Tax-A   0.25  -16.000000000   -4.000000000  -20.000000000']]
+    else:
+        assert get_pretty_tax_summary(source) == [
+            'Tax-A   0.25  180.000000000   45.000000000  225.000000000']
+        assert get_pretty_line_taxes_of_discount_lines(source) == [[
+            'Tax-A   0.25  -20.000000000   -5.000000000  -25.000000000']]
```

Comment:
Based on this. Should the line total be -20. If merchant use taxless prices in shop, is the only way to set discount amount 20 to actually calculate possible taxes. Sould the discount amount in campaings follow shop prices whether is taxless or taxfull.